File: Algorithms/0792_Number_of_Matching_Subsequences/Python/Number_of_Matching_Subsequences_Solution_2.py

# Space: O(n)
# Time: O(n)

import collections


class Solution:
    def numMatchingSubseq(self, s, words):
        res = 0
        self.cache = collections.defaultdict(list)
        self.result = {}
        for i in range(len(s)):
            self.cache[s[i]].append(i)
        for word in words:
            if word in self.result:
                res += self.result[word]
            else:
                if self.is_subsequence(word):
                    res += 1
        return res

    def is_subsequence(self, s):
        cur_index = -1
        for i in range(len(s)):
            if s[i] not in self.cache: return False
            temp = self.cache[s[i]]
            for j in temp:
                if j > cur_index:
                    cur_index = j
                    break
            else:
                self.result[s] = 0
                return False
        self.result[s] = 1
        return True

File: Algorithms/0792_Number_of_Matching_Subsequences/Java/NumberofMatchingSubsequencesSolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    private Map<Character, List<Integer>> cache = new HashMap<>();
    private Map<String, Integer> result = new HashMap<>();

    public int numMatchingSubseq(String s, String[] words) {
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            if (!cache.containsKey(s.charAt(i))) {
                cache.put(s.charAt(i), new ArrayList<>(Arrays.asList(i)));
            } else {
                cache.get(s.charAt(i)).add(i);
            }
        }
        for (String word : words) {
            if (result.containsKey(word)) {
                res += result.get(word);
                continue;
            }
            int tempRes = isSubsequence(word) ? 1 : 0;
            result.put(word, tempRes);
            res += tempRes;
        }
        return res;
    }

    private boolean isSubsequence(String s) {
        int tempIndex = -1;
        for (int i = 0; i < s.length(); i++) {
            if (!cache.containsKey(s.charAt(i))) {
                return false;
            }
            boolean flag = false;
            for (Integer index : cache.get(s.charAt(i))) {
                if (index > tempIndex) {
                    tempIndex = index;
                    flag = true;
                    break;
                }
            }
            if (!flag) return false;
        }
        return true;
    }
}




